modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java

```diff
@@ -17,94 +17,128 @@
-        this.snapshotStorage = snapshotStorage;
+    public IncomingSnapshotCopier(PartitionSnapshotStorage partitionSnapshotStorage, SnapshotUri snapshotUri) {
+        this.partitionSnapshotStorage = partitionSnapshotStorage;
         this.snapshotUri = snapshotUri;
     }
 
     @Override
-    public void cancel() {
-        //TODO https://issues.apache.org/jira/browse/IGNITE-17262
-        // Implement.
+    public void start() {
+        runAsync(this::start0, partitionSnapshotStorage.getIncomingSnapshotsExecutor());
     }
 
     @Override
     public void join() throws InterruptedException {
-        //TODO https://issues.apache.org/jira/browse/IGNITE-17262
-        // Implement proper join.
+        try {
+            future.get();
+        } catch (CancellationException e) {
+            LOG.info("Copier is cancelled");
+
+            if (!isOk()) {
+                setError(RaftError.ECANCELED, "Copier has been cancelled");
+            }
+        } catch (ExecutionException e) {
+            Throwable cause = e.getCause();
+
+            LOG.error("Error when completing the copier", cause);
+
+            if (cause instanceof SnapshotCopierException) {
+                SnapshotCopierException snapshotCopierException = (SnapshotCopierException) cause;
+
+                if (!isOk()) {
+                    setError(snapshotCopierException.status.getRaftError(), snapshotCopierException.status.getErrorMsg());
+                }
+
+                cause = snapshotCopierException.getCause();
+            } else {
+                if (!isOk()) {
+                    setError(RaftError.UNKNOWN, "Unknown error on completion the copier");
+                }
+            }
+
+            throw new IllegalStateException(cause);
+        }
     }
 
     @Override
-    public void start() {
-        //TODO https://issues.apache.org/jira/browse/IGNITE-17262
-        // What if node can't be resolved?
-        ClusterNode sourceNode = snapshotStorage.topologyService().getByConsistentId(snapshotUri.nodeName);
-
-        MessagingService messagingService = snapshotStorage.outgoingSnapshotsManager().messagingService();
+    public void cancel() {
+        cancelBusyLock.block();
 
-        threadPool.submit(() -> {
-            //TODO https://issues.apache.org/jira/browse/IGNITE-17262
-            // Following code is just an example of what I expect and shouldn't be considered a template.
-            CompletableFuture<NetworkMessage> metaRequestFuture = messagingService.invoke(
-                    sourceNode,
-                    MSG_FACTORY.snapshotMetaRequest().id(snapshotUri.snapshotId).build(),
-                    1000L
-            );
+        future.cancel(true);
+    }
 
-            metaRequestFuture.whenComplete((networkMessage, throwable) -> {
-                SnapshotMetaResponse metaResponse = (SnapshotMetaResponse) networkMessage;
+    @Override
+    public void close() throws IOException {
+        cancel();
```

Comment:
Shouldn't this be called when copier already canceled or joined? Very suspicious

Reply:
What do you mean by "tried to fix it"? `cancel()` call is still there. Is this really required? Please add a comment with a reason